src/transformers/modeling_utils.py

```diff
@@ -106,12 +106,15 @@ def no_init_weights(_enable=True):
     TODO(Patrick): Delete safety argument `_enable=True` at next major version. .
     """
     global _init_weights
-    if _enable:
-        _init_weights = False
+    old_init_weights = _init_weights
+    value = False if _enable else _init_weights
+    _init_weights = value
     try:
         yield
     finally:
-        _init_weights = True
+        if _init_weights != value:
+            logger.warning("_init_weights was changed inside no_init_weights context")
```

Comment:
Do we really need this warning? I have a hard time making sense of how does this help the user?

to me this looks like a debug log at best, and since your PR restores the original it's not even true any longer.

Does it make sense?

Reply:
The motivation of this warning is that, when user change the global inside this context(since python doesn't really have private), they might expect their change will be valid after exit this context. This warning can inform the restoration which might be forgotten by users.
Do you think that make sense?

Reply:
It'd probably help to share a use-case of where it'd happen. 

My guess is that you are trying to protect from code running in the `no_init_weights` context changing this global value, correct? Should there be an assert instead then as practically this situation breaks the context.

But if we add this warning, it'd be more informative not just say the value was changed, but what it was changed to?

Reply:
Actually I'd like to use an assert, if you don't think it is too aggressive.
The only way that can change `_init_weights` (without restoration) is access this private variable directly, which doesn't have any compatibility guarantee. Thus it is legal to use an assert.

Reply:
Let's not spend a lot of our time wondering what would happen if users change private variables. Even if nothing is truly private in Python, it's marked with an underscore so clearly indicated private and user that touch it do so at their own peril.

Therefore, let's not bother ourselves with that situation and remove the `value` variable as well as this warning.

Reply:
That sounds good to me, Sylvain.

Reply:
Sounds good!